rlpr-2.06/man/rlpr.1
.\" SCCS "@(#)rlpr.1	1.3	01/01/02 meem"
.TH rlpr 1 "2000/12/30" "rlpr 2.04" "UNIX Reference Manual"
.SH NAME
rlpr \- remote off-line print
.SH SYNOPSIS
\fBrlpr\fP [\fB-H\fPprinthost] [\fB-P\fPprinter] [\fB-X\fPproxyhost]
[\fB-#\fPcopies] [\fB-C\fPclass] [\fB-J\fPjob] [\fB-T\fPtitle] [\fB-U\fPuser]
[\fB-i\fP[indent]] [\fB-w\fPwidth] [\fB-1234cdfghlmnopqrstvNV\fP]
[\fB--debug\fP]
[\fB--port\fP=port]
[\fB--hostname\fP=hostname]
[\fB--send-data-first\fP]
[\fB--timeout\fP=seconds]
[\fB--tmpdir\fP=dir]
[\fB--verbose\fP]
[\fB--windows\fP]
[file ...]
.SH DESCRIPTION
\fBrlpr\fP uses TCP/IP to send print jobs to lpd servers anywhere on
a network.  Unlike \fBlpr\fP, it does not require that the remote printers
be explicitly known to the local system (traditionally through
\fI/etc/printcap\fR), and thus is considerably more flexible and
requires less administration.
.PP
\fBrlpr\fP can be used anywhere a traditional \fBlpr\fP might be used,
and is backwards compatible with traditional BSD \fBlpr\fP.  If
\fBrlpr\fP is invoked as \fBlpr\fP, it preserves all known lpr
semantics.
.PP
\fBrlpr\fP can be installed and used in two different ways:  (the
same guidelines apply for \fBrlpq\fP(1) and \fBrlprm\fP(1))
.PP
1. Ideally, if the \fBrlpr\fP client can be installed setuid root,
then \fBrlpr\fP can interact directly with other \fBlpd\fP's on the
network.  See \fBEXAMPLES\fP below.
.PP
2. If \fBrlpr\fP cannot be installed setuid root (in cases where one
is not the administrator on the machine one needs to print from) then
\fBrlpr\fP can be used in conjunction with an \fBrlprd\fP proxy
agent. See  \fPCONFIGURING A PROXY\fP and \fB EXAMPLES\fP below
for details.
.SH MOTIVATION
Traditional Berkeley \fBlpr\fP has one major drawback: the mortal user
cannot \fBlpr\fP to a printer which is not listed explicitly in the
\fI/etc/printcap\fR file.
.PP		
This is especially annoying in scenarios where there is a central UNIX
server which many users (using their own desktop computers) connect to
in order to do their daily work; If those users wish to print files
from the central server to their personal machines (assuming they had
a local \fBlpd\fP), they \fIcannot do this\fR without the central
server's \fI/etc/printcap\fR being edited to contain the host and
printer of each user on the network \- this scales horribly.
.SH OPTIONS
.PP
.SS Environment/Setup
.TP
.I      \-\-debug
Print gobs of debugging information.
.TP
.I \-N, \-\-no-bind
Don't try to bind to privileged port before connecting to lpd.  Please 
see \fBCONFIGURING A PROXY\fP below for information on use 
of this option.
.TP
.I      \-\-port=number
Select an alternate port (instead of 7290) to connect to, if using
rlprd.  Usually not needed.
.TP
.I \-H, \-\-printhost=host
Select the host to print to (used with \fI-P\fR).

Instead of using \fI-H\fR, one can specify the hostname directly
including it with the printer name with the \fIprinter@hostname\fR
syntax.
.TP
.I \-P, \-\-printer=printername, \-\-queue=printername
Select the printer to print to (used with \fI-H\fR).
.TP
.I \-X, \-\-proxyhost=proxyhost
Select the proxy host to use, if necessary.
.TP
.I \-\-timeout=seconds
Set the inactivity timer.  If the connection hangs for more than
\fBseconds\fP seconds, then \fBrlpr\fP will give up.  Use the special
value `-1' to wait forever.  Default timeout is 3 seconds.
.TP
.I \-\-tmpdir=tmpdir
Use tmpdir for temporary files (by default, \fB/tmp\fP is used)
.TP
.I \-V, \-\-version
Print version and exit.
.PP
.fi
\fBNOTE:\fP instead of specifying the printer, printhost, or proxy
variables on the commandline, one can set up defaults through either
environment variables or \fBrlprrc\fP configuration files.  For
information on personal or system-wide \fBrlprrc\fP files, see
\fBrlprrc\fP(5).
.SS Behavior
.TP
.I \-#, \-\-copies=copies
Print \fIcopies\fR copies of each document \- use sparingly,
\fBprinters aren't copiers\fP.
.TP
.I \-m, \-\-mail
Send mail upon completion.
.TP
.I \-q, \-\-quiet, \-\-silent
Quiet mode \-\ stay quiet (except for fatal errors).
.TP
.I \-r, \-\-remove
Remove file after printing.
.TP
.I \-s
.br
Ignored (provided for compatibility).
.TP
.I \-\-send\-data\-first
Send the "data" transaction before the "control" transaction.  Should
never be required when \fBrlpr\fP is used in conjunction with a RFC
1179 conformant \fBlpd\fP.  However, buggy implementations exist which 
require these semantics.  Use only when necessary.
.TP
.I \-\-verbose
Verbose mode \- on by default unless \fBrlpr\fP is invoked as \fBlpr\fP.
.TP
.I \-\-windows
"Windows" mode \- for printing multiple files to a
windows-based \fBlpd\fP.
.fi
.SS Document Content
.ta \w'1234'u
.nf
\fI\-1\fR	 Print data using \fBtroff\fP(1) R (times roman) font.
\fI\-2\fR	 Print data using \fBtroff\fP(1) I (times italic) font.
\fI\-3\fR	 Print data using \fBtroff\fP(1) B (times bold) font.
\fI\-4\fR	 Print data using \fBtroff\fP(1) S (special) font.
\fI\-c\fR	 Assume data has been produced by \fBcifplot\fP(1).
\fI\-d\fR	 Assume data has been produced by \fBtex\fP(1).
\fI\-f\fR	 Filter data assuming the start of each line has
	a \fIfortran\fR carriage control character.
\fI\-g\fR	 Assume data has been produced by the BSD plot library.
\fI\-l\fR	 Treat control characters as regular characters.
\fI\-n\fR	 Assume data has been produced by \fBditroff\fP.
\fI\-o\fR	 Assume data is \fIpostscript\fR.
\fI\-p\fR	 Print data using \fBpr\fP(1).
\fI\-t\fR	 Assume data has been produced by \fBtroff\fP(1).
\fI\-v\fR	 Assume the data contains a raster image.
.fi
.SS Document Format
.TP
.I \-i, \-\-indent=[cols]
Indent output by \fIcols\fR columns (8 by default).
.TP
.I \-h, \-\-no-burst
Do not print burst (banner) page.
.TP
.I \-w, \-\-width=width
Use \fIwidth\fR for page width when using \fBpr\fP(1).
.TP
.I \-C, \-\-class=class
Use \fIclass\fR for the class name on burst page instead of
the local system name.
.TP
.I \-J,  \-\-job=jobname
Use \fIjobname\fR for the job name on burst page instead of
the first filename in the job.
.TP
.I \-T, \-\-title=title
Use \fItitle\fR for the title when using \fBpr\fP(1)
instead of the current filename.
.TP
.I \-U, \-\-user=username
Use \fIusername\fR for the user name on burst page instead
of the actual username.
.TP
.I \-\-hostname=hostname
Use \fIhostname\fR for the hostname on the burst page instead of
the local system name.
.fi
.SS OS-Specific Extensions
See \fBOS-SPECIFIC EXTENSIONS\fR below for details on these
options.
.TP
.I \-\-ext=os \-\-extension=os
Interpret the arguments passed to \fI\-\-extargs\fR as if on operating 
system \fIos\fR.  Currently, \fIos\fR must be either \fIhpux\fR or
\fInone\fR.
.TP
.I \-\-extargs="args"
Change behavior according to OS-specific arguments listed in
\fIargs\fR.
.br
.SH "CONFIGURING A PROXY"
In situations where the \fBrlpr\fP client cannot be installed setuid
root, \fBrlpr\fP often cannot directly talk to an \fBlpd\fP because
most \fBlpd\fP's require that requests come from privileged
ports. Unfortunately, \fBrlpr\fP cannot "come from a privileged port"
without having superuser status.
.br
\fBThe privileged port requirement is part of RFC 1179.\fP

That said, some \fBlpd\fP's, most notably some of the ones which have
shipped with \fBWindows NT\fP, \fBSolaris\fP, and \fBNCRunix\fP, do
not require requests to come from a privileged port.  Before going
through the work of configuring a proxy, you should make sure one is
required for your situation.  You can check by trying to run
\fBrlpr\fP: if it is successful in spooling the job, you do not need
to install \fBrlpr\fP setuid root.  You should then pass the \fI\-N\fR
(or the more verbose \fI\-\-no-bind\fR) to \fBrlpr\fP in the future to
suppress \fBrlpr\fP's warning messages.

\fBIf your situation does not require a privileged port, do not use
one.\fP You can explicitly request a regular port, even when
\fBrlpr\fP has been installed setuid root, by using the
\fI\-\-no-bind\fR option.  This is a good idea because there are only
11 privileged ports that the RFC allows.  This ends up having the
practical implication of limiting your print requests to about 11 per
every 3 minutes, which is often unacceptable.

The rest of this section assumes that \fBrlpr\fP cannot be installed
setuid root in your situation, and that a remote \fBlpd\fP requires
\fBrlpr\fP connect to it from a privileged port.  It requires that you
or someone you know has root access to another machine the network, so
that you can use a \fBproxy\fP, known as \fBrlprd\fP, to get around
the problem.
.PP
\fBrlprd\fP works by taking \fBlpd\fP requests on a non-privileged port,
mapping them to a privileged port, and sending them out to the real
\fBlpd\fP. The proxy and the lpd \fBneed not be on the same
machine\fP. 
.PP
In the scenario provided in the motivation, if each user had root
access on his own machine, it would be optimal for the user to start
\fBrlprd\fP up on his own machine, in which case the proxy and the
printhost are the same machine.  The user would additionally have to
make sure his own machine was listed in his local
\fI/etc/hosts.lpd\fR so that his \fBlpd\fP would accept connections
from his \fBrlprd\fP.
.PP
However, if the machine the user wants to print to is not running
unix (but is running an \fBlpd\fP) (i.e. Windows, WinNT, etc) -- Then
that user can configure his \fBlpd\fP to accept connections from
an \fBrlprd\fP running somewhere on the network, and use that
\fBrlprd\fP as a proxy for \fBrlpr\fP.
.PP
For more on this mess, see \fBrlprd\fP(8).
.SH EXAMPLES
.TP
WITHOUT A PROXY  (rlpr is setuid root)
.PP
sun% \fBrlpr --printer=hp4l@foo.bar.org foo.c\fP
.PP
prints \fBfoo.c\fP to the printer \fBhp4l\fP on the host
\fBfoo.bar.org\fP (assuming foo.bar.org has been
configured to accept print requests from host \fBsun\fP in its
\fI/etc/hosts.lpd\fR).
.PP
sun% \fBwho | rlpr --printer=hp4l@foo.bar.org\fP
.PP
prints the output of \fBwho\fP(1) to the printer \fBhp4l\fP on the
host \fBfoo.bar.org\fP.
.TP
WITH A PROXY  (rlpr is a normal program)
.PP
sun% \fBrlpr -Xfoo.bar.org --printer=hp4l@foo.bar.org foo.c\fP
.PP
prints \fBfoo.c\fP to the printer \fBhp4l\fP on the host
\fBfoo.bar.org\fP using \fBfoo.bar.org\fP as a proxyhost. Under this
configuration, foo.bar.org needs to be configured to accept print
requests from the proxy (itself) and \fBnot from sun\fP. In situations
where the proxy and the printhost are not the same machine, this can
lead to some security holes due to the original lameness of berkeley
\fBlpd\fP's host-based security.
.PP
\fBNOTE:\fP if a proxyhost isn't specified anywhere, \fBrlpr\fP
assumes one isn't necessary.
.SH "OS-SPECIFIC EXTENSIONS"
Some vendors have added features to their BSD-based \fBlpr\fP
products, which are upwardly compatible with \fBlpr\fP but
incompatible with one another.  To support these extensions,
\fBrlpr\fP allows the user to indicate what operating system to
emulate and provide a string of OS-specific options, either through 
commandline arguments or environment variables.
.PP
Note that for correct operation, these extensions should be only used
when interoperating with an \fBlpd\fP provided by the vendor for the
indicated operating system.
.SH "ENVIRONMENT VARIABLES"

rlpr-2.06/NEWS
Changes in release 2.06:

* fixed several autoconf-related problems introduced in 2.05.
* support fallback to syslog() when vsyslog() is not available.

Changes in release 2.05:

* fixed security flaw detailed at: http://www.osvdb.org/7194

Changes in release 2.04:

* removed -W flag so that it does not conflict with POSIX.2 systems.
* fixed rlpr to not expect an argument with --windows (nee -W) option.
* fixed `bsdlinks' target to not accidentally remove lprm.
* stubbed out i18n/l10n functions to make ./configure --disable-nls work.
* enhanced rlpr to print a warning if the user attempts to print an
  empty file.
  
Changes in release 2.03:

* added support for printer alternation in check-server.sh.
* fixed message displayed when incorrect non-long style options are
  passed to rlpr, rlpq, rlprd, or rlprm.
* added configurable configuration directory (RLPR_CONFDIR) support.
* added support for running `configure' out of other than `top_srcdir'
* fixed build problem that caused libnsl and libsocket (when required)
  not to appear on link line.
* new autoconf macros for EXIT_FAILURE/EXIT_SUCCESS.
* fixed two bugs in rlprrc lookups of hosts for a printer.
* fixed bug which prevented rlprrc lookups for a printer for a host
  (now, default to `lp' is only done when the rlprrc lookup fails).

Changes in release 2.02:

* fixed connect() to report the correct error message upon failure.
* changed Makefiles to never use += operator (for portability).
* revised code to compile correctly on newer releases of GNU libc.
* updated pointers to packages in README.
* fixed --no-bind to work when running as root.
* fixed Makefiles so error.o is now only put in librlpr once.
* fixed broken configure.in custom checks for MAXHOSTNAMELEN and off_t.
* added custom check for h_errno definition in <netdb.h>.
* fixed test(1) brokenness in check-server.sh.

Changes in release 2.01:

* fixed bug that would cause rlpr to send from unprivileged ports
* fixed some longstanding argument parsing problems
* added `--hostname' option
* fixed bug that caused rlpr to be interrupted with SIGPOLL
* added generic support for vendor extensions to lpr
* added support for some hp/ux lpr extensions
* changed the default timeout to be more reasonable
* small fixes to make the code compile properly on aix 4.1.x
* small fixes to make the code compile properly on freebsd
* made ~/.rlprrc truly optional

Changes in release 2.00:

* complete rewrite of rlpr, rlpq, and rlprd - much more modular
* addition of new rlprm command
* complete re-architecture of build environment (deep directories now)
* new hooks for internationalization support
* build procedure now uses automake in addition to autoconf
* remote printing commands now support a --timeout option in case
  remote end dies
* new --send-data-first option to rlpr to work with buggy lpd's
* rlpr now works with hp jetdirect printers
* rlpr now includes a shiny-new progress meter, useful when printing
  files over slow links
* all printing commands default to `lp' printer now, like bsd
* RLPR_PROXY has been removed; use RLPR_PROXYHOST instead
* uses automake, autoconf, etc..


rlpr-2.06/src/rlpr.c
/*
 * Copyright (c) 1998-1999 peter memishian (meem), meem@gnu.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

#pragma ident "@(#)rlpr.c	1.2	01/01/01 meem"

#include <config.h>
#include <stdlib.h>
#include <stdio.h>
#include <locale.h>
#include <getopt.h>
#include <errno.h>
#include <unistd.h>
#include <pwd.h>
#include <string.h>
#include <sys/utsname.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <lib.h>
#include R_MAXHOSTNAMELEN_HDR

#include "intl.h"
#include "component.h"
#include "msg.h"
#include "util.h"
#include "client.h"
#include "rlprrc.h"
#include "rlpr.h"
#include "rfc1179.h"

static struct component *components[] =
{
    &comp_msg,
    &comp_component,
    &comp_client,
    &comp_rlpr,
    0
};

static struct rlpr_rlpr	*rlpr_rlpr;

static char	*finished_fmt[] = {
    "%i files spooled to %s@%s (proxy %s)",
    "%i file spooled to %s@%s (proxy %s)"
};

const char	*program_name;
const char	*bsd_program_name = "lpr";

static int	send_controlf(int, int, const char *);
static int	send_dataf(int, const char *, const char *);
static void	controlf_add(int, char, const char *, const char *);
static void	controlf_init(int, const char *);
static void	tmp_file_next(int, char *, char *, char *);
static int	tmp_file_init(char **, char **, char **);

int
main(int argc, char *argv[])
{
    struct component	*comp;
    char		*cf_name, *df_name, *cf_path;
    char		*req;
    int			 cf_fd, sock_fd = -1;
    unsigned int	 n_files, files_left, n_copies;

    program_name = argv[0];
    toggle_root();

    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);

    if ((comp = component_init(components, argc, argv)) != 0)
	msg(R_FATAL, 0, "component `%s': init() failed!", comp->name);

    /*
     * create the name templates for control and data files, and the files..
     */

    cf_fd = tmp_file_init(&cf_path, &cf_name, &df_name);
    if (cf_fd == -1)
	msg(R_FATAL, 0, "cannot create temporary files");

    n_files = argc - optind;
    if (n_files == 0)		/* it's from stdin */
	n_files++;

    /*
     * create the job request.  nasty.
     */

    req = xmalloc(R_REQ_LEN);
    sprintf(req, "%c%.*s\n", RECVJ, R_REQ_LEN - 5, client_get_printer());

    /*
     * now print each file.  note that we have to continuously
     * re-initialize the connection when connecting to windows lpds.
     */

    for (files_left = n_files; files_left != 0; files_left--, optind++) {

	if (sock_fd == -1 || rlpr_rlpr->windows_lpd) {

	    sock_fd = client_open(rlpr_rlpr->timeout);
	    if (sock_fd == -1)
		msg(R_FATAL, 0, "client_open(): cannot connect to lpd");

	    if (client_command(sock_fd, req, rlpr_rlpr->timeout, "job request")
		== 0)
		msg(R_FATAL, 0, "unable to send job request to lpd");
	}

	controlf_init(cf_fd, argv[optind] ? argv[optind] : "stdin");

	for (n_copies = 0; n_copies < rlpr_rlpr->n_copies; n_copies++)
	    controlf_add(cf_fd, rlpr_rlpr->filetype, df_name, 0);

	controlf_add(cf_fd, 'U', df_name, 0);
	controlf_add(cf_fd, 'N', argv[optind], "stdin");

	/*
	 * make sure we can access the file here, before sending the
	 * control file.
	 */

	if (argv[optind] != 0 && access(argv[optind], R_OK) == -1) {

	    msg(R_ERROR, errno, "skipping file `%s'", argv[optind]);
	    n_files--;

	} else {

	    if (rlpr_rlpr->control_first)
		if (send_controlf(sock_fd, cf_fd, cf_name) == 0)
		    msg(R_FATAL, 0, "unable to control file to lpd");

	    if (send_dataf(sock_fd, argv[optind], df_name) == 0)
		msg(R_FATAL, 0, "unable to send file `%s' to lpd",
		    argv[optind] ? argv[optind] : "stdin");

	    if (rlpr_rlpr->control_first == 0)
		if (send_controlf(sock_fd, cf_fd, cf_name) == 0)
		    msg(R_FATAL, 0, "unable to control file to lpd");

	    if (rlpr_rlpr->remove_after && argv[optind])
		if (unlink(argv[optind]) == -1)
		    msg(R_WARNING, errno, "cannot remove `%s'", argv[optind]);
	}

	if (rlpr_rlpr->windows_lpd)
	    close(sock_fd);

	tmp_file_next(cf_fd, cf_path, cf_name, df_name);
    }

    msg(R_INFO, 0, finished_fmt[n_files == 1], n_files,
	client_get_printer(), client_get_printhost(), client_get_proxyhost());

    if ((comp = component_fini(components)) != 0)
	msg(R_FATAL, 0, "component `%s': fini() failed!", comp->name);

    return EXIT_SUCCESS;
}

int
rlpr_init(void)
{
    struct rlpr_rlpr	rlpr_rlpr_tmpl = { 0 };
    struct utsname      utsname;
    struct passwd      *pwd;

    /*
     * yeah, i know this same code is in client_init().  however,
     * there is a method to this madness: their uses of localhost
     * are not tied to each other, and thus neither should their
     * setting of it.
     */

    if (uname(&utsname) == -1) {
	msg(R_ERROR, errno, "uname failed");
	return 0;
    }

    /*
     * we perform a structure copy on a zero-initialized structure
     * instead of using memset() since memset is a raw bit-oriented
     * function, which means that it will generate potentially
     * incorrect results with non-integer types.
     */

    rlpr_rlpr	= xmalloc(sizeof (struct rlpr_rlpr));
    *rlpr_rlpr	= rlpr_rlpr_tmpl;

    /*
     * explicitly initialize all non-zero parameters
     */

    rlpr_rlpr->timeout  = R_TIMEOUT_DEFAULT;
    rlpr_rlpr->n_copies	= 1;
    rlpr_rlpr->filetype = 'f';
    rlpr_rlpr->print_burst++;
    rlpr_rlpr->control_first++;

    pwd = getpwuid(getuid());
    if (pwd == 0 || pwd->pw_name == 0) {
	msg(R_ERROR, errno, "unable to resolve your username");
	return 0;
    }

    rlpr_rlpr->localhost = xstrdup(utsname.nodename);
    rlpr_rlpr->user	 = xstrdup(pwd->pw_name);
    rlpr_rlpr->tmpdir	 = getenv("TMPDIR");
    if (rlpr_rlpr->tmpdir == 0)
	rlpr_rlpr->tmpdir = R_TMPDIR;

    return 1;
}

static int
rlpr_parse_args(int opt)
{
    switch (opt) {

    case '1':		/* troff type R */
    case '2':		/* troff type I */
    case '3':		/* troff type B */
    case '4':		/* troff type S */
    case 'c':		/* cifplot */
    case 'd':		/* TeX/DVI */
    case 'f':		/* fortran */
    case 'g':		/* graph */
    case 'l':		/* leave control characters */
    case 'n':		/* ditroff */
    case 'o':		/* postscript (not in lpr) */
    case 'p':		/* pr */
    case 't':		/* troff */
    case 'v':		/* raster input */

	rlpr_rlpr->filetype = (opt == 'f') ? 'r' : (char)opt;
	break;

    case '#':
	rlpr_rlpr->n_copies = strtoul(optarg, 0, 0);
	break;

    case 'h':
	rlpr_rlpr->print_burst = 0;
	break;

    case 'm':
	rlpr_rlpr->mail_after = 1;
	break;

    case 'r':
	rlpr_rlpr->remove_after = 1;
	break;

    case -405:
	rlpr_rlpr->windows_lpd = 1;
	break;

    case 's':		/* bsd compatibility */
	msg(R_WARNING, 0, "symlink option not applicable (ignored)");
	break;

    case 'C':
	rlpr_rlpr->class = optarg;
	break;

    case 'J':
	rlpr_rlpr->job = optarg;
	break;

    case 'T':
	rlpr_rlpr->title = optarg;
	break;

    case 'U':
	rlpr_rlpr->user = optarg;
	break;

    case -404:
	rlpr_rlpr->tmpdir = optarg;
	break;

    case -401:
	msg(R_STDOUT, 0, "usage: %s [-Hprinthost] [-Pprinter] [-Xproxy]"
	    " [OPTIONS] [file ...]\nplease see the manpage for detailed"
	    " help", program_name);
	exit(EXIT_SUCCESS);
	break;

    case -400:
	rlpr_rlpr->control_first = 0;
	break;

    case -403:
	rlpr_rlpr->timeout = strtol(optarg, 0, 0);
	break;

    case -402:
	rlpr_rlpr->hostname = optarg;
	break;

    case 'V':
	msg(R_STDOUT, 0, "version "VERSION" from " __DATE__" "__TIME__
	    " -- meem@gnu.org");
	exit(EXIT_SUCCESS);
	break;

    case 'w':
	rlpr_rlpr->width = optarg ? optarg : R_NOARG_WIDTH;
	break;

    case 'i':
	rlpr_rlpr->indentation = optarg ? optarg : R_NOARG_INDENT;
	break;

    case EOF:
	break;
    }

    return 1;
}

static int
rlpr_hpux_parse_args(int opt)
{
    msg(R_DEBUG, 0, "optarg: %s", optarg);

    switch (opt) {

    case 'o':
	rlpr_rlpr->options = optarg;
	break;

    case 'p':
	rlpr_rlpr->priority = optarg;
	break;
    }

    /* XXX write_back isn't settable yet; need to check hpux for flag */
    return 1;
}

static off_t
send_header(int sock_fd, int fd, const char *name, char cmd, const char *what)
{
    off_t	fd_size = file_size(fd);
    char	hdr[R_BUFMAX];

    if (fd_size == (off_t)-1) {
	msg(R_ERROR, 0, "send_header: cannot determine size of %s", name);
	return -1;
    }

    sprintf(hdr, "%c"R_OFF_T_FMT" %.*s\n", cmd, fd_size, MAXHOSTNAMELEN, name);
    if (client_command(sock_fd, hdr, rlpr_rlpr->timeout, what) == 0)
	return -1;

    return fd_size;
}

static int
send_controlf(int sock_fd, int cf_fd, const char *cf_name)
{
    char	buffer[R_BUFMAX];
    ssize_t	n_bytes_sent = 0;
    FILE	*cfp;
    off_t	cf_size;

    /*
     * send the control file header
     */

    cf_size = send_header(sock_fd, cf_fd, cf_name, RECVCF, "control file hdr");
    if (cf_size == -1) {
	msg(R_ERROR, 0, "send_controlf: cannot send control file header");
	return 0;
    }

    /*
     * now send the control file over...
     */

    lseek(cf_fd, 0, SEEK_SET);
    cfp = fdopen(cf_fd, "r");
    if (cfp == 0) {
	msg(R_ERROR, errno, "send_controlf: fdopen");
	return 0;
    }

    while (fgets(buffer, sizeof buffer, cfp) != 0) {

	size_t length = strlen(buffer);

	buffer[length - 1] = '\0';
	msg(R_DEBUG, 0, "control file: %s", buffer);
	buffer[length - 1] = '\n';

	if (full_write_timed(sock_fd, buffer, length, rlpr_rlpr->timeout, 0, 0)
	    < length) {
	    msg(R_ERROR, 0, "send_controlf: timed out sending control file");
	    return 0;
	}

	n_bytes_sent += length;
	msg(R_TTY, 0, "control file: sent %d%\r", n_bytes_sent * 100 / cf_size);
    }

    if (full_write_timed(sock_fd, "", 1, rlpr_rlpr->timeout, 0, 0) != 1) {
	msg(R_ERROR, 0, "unable to send control file to lpd");
	return 0;
    }

    return check_ack(sock_fd, "control file contents", rlpr_rlpr->timeout) == 1;
}

static int
send_dataf(int sock_fd, const char *filename, const char *df_name)
{
    int		df_fd;
    off_t	df_size;
    char	*t_filename;

    /*
     * the rfc states tha a data file size of 0 denotes a data file of
     * unspecified size (which is useful when rlpr is streaming from
     * stdin).  however, it appears some lpd's can't cope with this, so we
     * always create a temporary file when reading from stdin.  grrr...
     */

    if (filename != 0) {

	df_fd = open(filename, O_RDONLY);
	if (df_fd == -1) {
	    msg(R_ERROR, errno, "send_dataf: open on `%s'", filename);
	    return 0;
	}

    } else {

	filename   = "stdin";
	t_filename = xmalloc(strlen(rlpr_rlpr->tmpdir) + strlen(df_name) + 2);
	sprintf(t_filename, "%s/%s", rlpr_rlpr->tmpdir, df_name);

	df_fd = open(t_filename, O_RDWR|O_CREAT|O_EXCL, 0600);
	if (df_fd == -1) {
	    msg(R_ERROR, errno, "send_dataf: cannot create temporary datafile");
	    return 0;
	}

	unlink(t_filename);
	free(t_filename);

	if (copy_file(STDIN_FILENO, df_fd, -1, 0, "temporary datafile") == 0) {
	    msg(R_ERROR, 0, "unable to create temporary copy of stdin for lpd");
	    return 0;
	}

	lseek(df_fd, SEEK_SET, 0);
    }

    df_size = send_header(sock_fd, df_fd, df_name, RECVDF, "data file hdr");
    if (df_size == -1) {
	msg(R_ERROR, 0, "send_dataf: sending data file header failed");
	return 0;
    }

    if (df_size == 0)
	msg(R_WARNING, 0, "%s is an empty file", filename);

    if (copy_file(df_fd, sock_fd, rlpr_rlpr->timeout, df_size, filename) == 0) {
	msg(R_ERROR, 0, "unable to send data file to lpd");
	return 0;
    }

    if (full_write_timed(sock_fd, "", 1, rlpr_rlpr->timeout, 0, 0) != 1) {
	msg(R_ERROR, 0, "unable to send data file to lpd");
	return 0;
    }

    return check_ack(sock_fd, "data file contents", rlpr_rlpr->timeout) == 1;
}

static void
controlf_add(int cf_fd, char op, const char *value, const char *def)
{
    char	*entry;

    if (value == 0)
	value = def;

    msg(R_DEBUG, 0, "value %s, default %s", value == 0 ? "(null)" : value,
	def == 0 ? "(null)" : def);

    if (value != 0) {
	entry = xmalloc(strlen(value) + 3);
	sprintf(entry, "%c%s\n", op, value);
	msg(R_DEBUG, 0, "wrote %c%s", op, value);
	if (full_write(cf_fd, entry, strlen(entry), 0) == 0)
	    msg(R_WARNING, errno, "controlf_add: full_write");
	free(entry);
    }
}

static void
controlf_init(int cf_fd, const char *file_to_print)
{
    if (ftruncate(cf_fd, 0) == -1)
	msg(R_WARNING, 0, "controlf_init: ftruncate");

    controlf_add(cf_fd, 'H', rlpr_rlpr->hostname, rlpr_rlpr->localhost);
    controlf_add(cf_fd, 'P', rlpr_rlpr->user, 0);
    controlf_add(cf_fd, 'I', rlpr_rlpr->indentation, 0);
    controlf_add(cf_fd, 'T', rlpr_rlpr->title, 0);

    if (rlpr_rlpr->mail_after)
	controlf_add(cf_fd, 'M', rlpr_rlpr->user, 0);

    if (rlpr_rlpr->print_burst) {

	/*
	 * the defaults on the 'J' and 'C' options are meant to
	 * appease berkeley lpd, which is apparently more finicky than
	 * the rfc calls for.
	 */

	controlf_add(cf_fd, 'J', rlpr_rlpr->job, file_to_print);
	controlf_add(cf_fd, 'C', rlpr_rlpr->class, rlpr_rlpr->localhost);
	controlf_add(cf_fd, 'L', rlpr_rlpr->user, 0);
    }

    /* another berkeleyism.  tolerate. */

    if (strchr("flp", rlpr_rlpr->filetype) != 0) {
	controlf_add(cf_fd, 'W', rlpr_rlpr->width, 0);
    }

    /*
     * handle any extensions
     */

    switch (component_get_extension()) {

    case R_EXT_HPUX:

	if (rlpr_rlpr->write_back)
	    controlf_add(cf_fd, 'R', rlpr_rlpr->user, 0);

	controlf_add(cf_fd, 'A', rlpr_rlpr->priority, 0);
	controlf_add(cf_fd, 'O', rlpr_rlpr->options, 0);
	break;
    }
}

static int
tmp_file_init(char **cf_path, char **cf_name, char **df_name)
{
    int		cf_fd;

    *cf_path = xmalloc(strlen(rlpr_rlpr->localhost) +
		       strlen(rlpr_rlpr->tmpdir) + sizeof "/cfAxxx");
    sprintf(*cf_path, "%s/cfA%03d", rlpr_rlpr->tmpdir, (int)getpid() % 1000);
    strcat(*cf_path, rlpr_rlpr->localhost);

    *cf_name = strrchr(*cf_path, '/') + 1;
    *df_name = xstrdup(*cf_name);
    *df_name[0] = 'd';

    cf_fd = open(*cf_path, O_RDWR|O_TRUNC|O_CREAT, 0600);
    if (cf_fd == -1) {
	msg(R_ERROR, errno, "tmp_file_init: open: %s", *cf_path);
	free(*df_name);
	free(*cf_path);
    }

    msg(R_DEBUG, 0, "tmp_file_init: cf_fd: %i (%s)", cf_fd, *cf_path);

    unlink(*cf_path);
    return cf_fd;
}

/* ARGSUSED */
static void
tmp_file_next(int cf_fd, char *cf_path, char *cf_name, char *df_name)
{
    cf_name[2]++;		/* cf_path shares the same pointer */
    df_name[2]++;

    lseek(cf_fd, 0, SEEK_SET);
}

static struct option rlpr_opts[] = {
    { "copies",			1, 0, '#'  },
    { "send-data-first",	0, 0, -400 },
    { "help",			0, 0, -401 },
    { "hostname",		1, 0, -402 },
    { "indent",			2, 0, 'i'  },
    { "job",			1, 0, 'J'  },
    { "mail",			0, 0, 'm'  },
    { "no-burst",		0, 0, 'h'  },
    { "remove",			0, 0, 'r'  },
    { "timeout",		1, 0, -403 },
    { "title",			1, 0, 'T'  },
    { "tmpdir",			1, 0, -404 },
    { "user",			1, 0, 'U'  },
    { "version",		0, 0, 'V'  },
    { "width",			2, 0, 'w'  },
    { "windows",		0, 0, -405 },
    { 0,			0, 0,  0   }
};

static const char rlpr_opt_list[] = "1234cdfglnoptvFsC:#:i::J:mhrT:U:w::W:V";
static const char rlpr_hpux_opt_list[] = "o:p:";

struct component comp_rlpr = {
    "rlpr",  rlpr_init, 0,
    { { rlpr_opts, rlpr_opt_list, rlpr_parse_args },
      { 0, rlpr_hpux_opt_list, rlpr_hpux_parse_args }
    }
};
